app/core/database.py

```diff
@@ -0,0 +1,68 @@
+import contextlib
+from typing import AsyncIterator
+
+from sqlalchemy.ext.asyncio import (
+    AsyncConnection,
+    AsyncEngine,
+    AsyncSession,
+    async_sessionmaker,
+    create_async_engine,
+)
+from sqlmodel import SQLModel
+
+
+class DatabaseSessionManager:
+    def __init__(self):
+        self._engine: AsyncEngine | None = None
+        self._sessionmaker: async_sessionmaker | None = None
+
+    def init(self, host: str):
+        self._engine = create_async_engine(host)
+        self._sessionmaker = async_sessionmaker(autocommit=False, bind=self._engine)
+
+    async def close(self):
+        if self._engine is None:
+            raise Exception("DatabaseSessionManager is not initialized")
+        await self._engine.dispose()
+        self._engine = None
+        self._sessionmaker = None
+
+    @contextlib.asynccontextmanager
+    async def connect(self) -> AsyncIterator[AsyncConnection]:
+        if self._engine is None:
+            raise Exception("DatabaseSessionManager is not initialized")
+
+        async with self._engine.begin() as connection:
+            try:
```

Comment:
My Python-fu is weak: this block confuses me - is it possible for `yield connection` to throw an exception? It looks to me like this try block is pointless. Or does the `except` also catch anything thrown by what follows `with` despite that not being within `try`?

Reply:
given that is marked with the asynccontextmanager decorator, it is used elsewhere e.g.
```python
async with contextmanager.connect() as connection:
   ... do stuff
```
if "do stuff" raises an exception, the except block within the context manager catches it and ensures the connection is rolled back so nothing is contaminated. 

Added long wordy comments to this entire class